Add Back "Classic Zoom Levels"

andrikpowell opened the pull request:
This adds back the functionality of being able to shrink/grow the screensize via "-"/"+".

I've also added a few improvements:
- fixed weapon interpolation causing the weapon to "wiggle" during screensize changes
- Added a menu option to enable or disable the feature. This means that players who don't like the classic zooms and prefer how dsda-doom is currently set up can continue to use dsda-doom without this feature (I've set the option to "no" by default)

There are two things I'd like to mention that I haven't been able to solve / figure out.
1. I'd like to scale the screensize border graphics similar to the grnrock flat
2. This is a carryover from PrBoom, but it's weird that the automap is forced in the screensize, instead of Vanilla where the automap is the same fullscreen size regardless of screen zoom size. I think it makes more sense to have the automap fullscreen at all times, as it's useless to use under smaller screen sizes.

fabiangreffrath commented:
I think it's redundant to add a toggle for a feature that requies explicit key presses to take action anyway. If players are afraid of accidentally changing the screen size, they can just unbind the keys, or not?

andrikpowell committed:
Add back screen zoom levels

andrikpowell commented:
> I think it's redundant to add a toggle for a feature that requies explicit key presses to take action anyway. If players are afraid of accidentally changing the screen size, they can just unbind the keys, or not?

I guess my logic was that the feature was removed for a reason, and perhaps the reason was that Kraflab didn't like the smaller screensizes? So I guess when I implemented the config option it was to allow players to have the screensize functionality as it currently works.

But if we think we don't need the option, I could scrap the option code pretty easily.

_Also ignore the force-push, I just cleaned up some of the code and removed some unneeded stuff (ex: the "paused" graphic had a "viewwindowy + 5" offset, but I tested it and it works the same way with or without, so I scrapped that edit)_

Pedro-Beirao commented:
I agree that we dont need an option for this. Just the keybinds.

andrikpowell committed:
Fix Weapon Interpolation when changing screensize

andrikpowell commented:
Alright y'all's wishes are my command :P

I'd rather not have the option if we don't need it, so I removed it from the branch. I can easily add it back if need be.

andrikpowell committed:
Merge branch 'master' into zoom-levels

rfomin commented:
> I guess my logic was that the feature was removed for a reason, and perhaps the reason was that Kraflab didn't like the smaller screensizes?

The reason was the same as for removing ENDOOM or PC Speaker emulation - remove useless code. Less code - less bugs and easier maintenance.

Have there been any complaints about the lack of this feature? Maybe we shouldn't restore it.

andrikpowell commented:
> The reason was the same as for removing ENDOOM or PC Speaker emulation - remove useless code. Less code - less bugs and easier maintenance.

I'm pretty sure the main reason the feature was cut was cuz at the time Kraflab completely redesigned the message / hud display... and alot of that logic was tied with the screensizes, for whatever reason.

So when I reimplemented the screensizes back, I didn't touch any of the message code that was removed... and it works fine with how the message code works currently.

> Have there been any complaints about the lack of this feature? Maybe we shouldn't restore it.

I'm definitely not a fan of just removing features that were present in the original exe. To be quite honest, I've thought about restoring PC Speaker emulation as well, but I'm sure that's an even less sought after feature. I realise I'm probably a bit weird about things like this, as one of the things that still bothers me is that there's no "end game" option in the main options menu.

But I think a feature like this is worth restoring as a majority of source ports already do, including even GZDoom. The only source port I can think of that doesn't currently support this is the KEX engine, which I personally disagree with, but that's a discussion for another time. I'm unsure if Helion does or not...

rfomin commented:
> But I think a feature like this is worth restoring as a majority of source ports already do, including even GZDoom. The only source port I can think of that doesn't currently support this is the KEX engine, which I personally disagree with, but that's a discussion for another time. I'm unsure if Helion does or not...

The KEX port supports the "screen size" feature, although it doesn't render the border. Helion doesn't support it at all. Personally, I do not think this feature is worth restoring. Not many users even know about it these days.

Another small issue - background flat tiling changes when the zoom level feature is activated.
Before:
![image](https://github.com/user-attachments/assets/b49e0746-b0d2-4175-9890-616144e9d5d7)
After:
![image](https://github.com/user-attachments/assets/ffeb9967-3a0c-4d1e-ae59-17666affcafc)

andrikpowell committed:
Make background flat tiling consistent (between screenviews)

andrikpowell commented:
> Another small issue - background flat tiling changes when the zoom level feature is activated.
> Before:
> ![image](https://private-user-images.githubusercontent.com/5077629/418402395-b49e0746-b0d2-4175-9890-616144e9d5d7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyNTUyODIsIm5iZiI6MTc0MTI1NDk4MiwicGF0aCI6Ii81MDc3NjI5LzQxODQwMjM5NS1iNDllMDc0Ni1iMGQyLTQxNzUtOTg5MC02MTYxNDRlOWQ1ZDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDZUMDk1NjIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGFlMmQxODhjZWY4NmM1YWYwODEyZjIzY2Q1YWQ5YzJkNGExNTE4MDk2ZGNmZjQyOTY2ZDQzZjFhZjUyZjA0MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.yghNOyaKijJd7VsHq5HpQVBOgmNTJvF_I2us7XjAh0k)
> After:
> ![image](https://private-user-images.githubusercontent.com/5077629/418402451-ffeb9967-3a0c-4d1e-ae59-17666affcafc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyNTUyODIsIm5iZiI6MTc0MTI1NDk4MiwicGF0aCI6Ii81MDc3NjI5LzQxODQwMjQ1MS1mZmViOTk2Ny0zYTBjLTRkMWUtYWU1OS0xNzY2NmFmZmNhZmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDZUMDk1NjIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTZlMWZhZjI5NWJhNWY3ZmY1OWM3ZDZhOWJhMmQyZTJhZDRlNTVmODk2NTA2YTRkYjg2NmRmMjhmZmMwMWI4ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.OTcJKJy4zEzErmxNb6pqz0eIJaR9pvygFwkelVWx8q4)

This should be fixed now.

Pedro-Beirao commented:
> Personally, I do not think this feature is worth restoring. Not many users even know about it these days.

Eh I agree. If it hadnt been removed, it wouldnt hurt to keep it, but at the same time, no point in bringing it back.

If for some reason there is demand for this in the future, we can merge this, but for now I think its fine to keep it a nyan only thing. Thank you for the PR anyway!

Pedro-Beirao commented:
And important to note that this was used to effectively reduce the resolution so that slower machines could play the game (3DO port is a good example), we dont need it now that we can choose the resolution.

So even though Im also not a fan of removing vanilla features, I see this as an improvement/replacement of the feature, not simply removing it.

Pedro-Beirao commented:
Closing as explained above
